src/lib.rs

```diff
@@ -283,50 +118,56 @@ impl Pty {
       cmd.args(args);
     }
 
-    let rustix_openpty::Pty {
-      controller: controller_fd,
-      user: user_fd,
-    } = openpty(None, Some(&window_size))
-      .map_err(|err| napi::Error::new(napi::Status::GenericFailure, err))?;
-
-    if let Ok(mut termios) = termios::tcgetattr(&controller_fd) {
-      termios.input_modes.set(InputModes::IUTF8, true);
-      termios::tcsetattr(&controller_fd, OptionalActions::Now, &termios)
-        .map_err(|err| napi::Error::new(napi::Status::GenericFailure, err))?;
-    }
+    // open pty pair
+    let pty_res = openpty(&window_size, None).map_err(cast_to_napi_error)?;
+    let controller_fd = pty_res.master;
+    let user_fd = pty_res.slave;
 
-    // The Drop implementation for Command will try to close _each_ stdio. That implies that it
-    // will try to close the three of them, so if we don't dup them, Rust will try to close the
-    // same FD three times, and if stars don't align, we might be even closing a different FD
-    // accidentally.
+    // duplicate pty user_fd to be the child's stdin, stdout, and stderr
     cmd.stdin(Stdio::from(user_fd.try_clone()?));
     cmd.stderr(Stdio::from(user_fd.try_clone()?));
     cmd.stdout(Stdio::from(user_fd.try_clone()?));
 
-    // We want the env to be clean, we can always pass in `process.env` if we want to.
+    // we want the env to be clean, we can always pass in `process.env` if we want to.
     cmd.env_clear();
     if let Some(envs) = opts.envs {
       cmd.envs(envs);
     }
+
+    // set working dir if applicable
     if let Some(dir) = opts.dir {
       cmd.current_dir(dir);
     }
 
     unsafe {
       let raw_user_fd = user_fd.as_raw_fd();
-      let raw_controller_fd = controller_fd.as_raw_fd();
+
+      // right before we spawn the child, we should do a bunch of setup
+      // this is all run in the context of the child process
       cmd.pre_exec(move || {
+        // start a new session
         let err = libc::setsid();
         if err == -1 {
-          return Err(Error::new(ErrorKind::Other, "Failed to set session id"));
+          return Err(Error::new(ErrorKind::Other, "setsid"));
         }
 
-        // stdin is wired to the tty, so we can use that for the controlling terminal.
-        set_controlling_terminal(0)?;
+        // become the controlling tty for the program
+        let err = libc::ioctl(raw_user_fd, libc::TIOCSCTTY.into(), 0);
+        if err == -1 {
+          return Err(Error::new(ErrorKind::Other, "ioctl-TIOCSCTTY"));
+        }
 
-        libc::close(raw_user_fd);
-        libc::close(raw_controller_fd);
+        // we dont need to drop any fds as we are the child and
+        // all fds we inherit have CLO_EXEC set
```

Comment:
uh, i don't think they do! they kinda sorta have to not have CLOEXEC to be part of the stdio set.

we're implicitly closing the userFD in L164, but we _are_ leaking the controller fd now.